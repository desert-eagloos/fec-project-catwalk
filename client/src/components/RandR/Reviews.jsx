import React from 'react';
import '../../css/RandR/Reviews.css';

const Reviews = () => (
  <div className="randr-reviews-container">
    <div>Total reviews and sorting</div>
    <div>
      <div>star rating</div>
      <div>user and date posted</div>
    </div>
    <div>review title</div>
    <div>review body</div>
    <div>helpful and report</div>
    <div>
      <button type="button">More Reviews</button>
      <button type="button">Add a Review</button>
    </div>
  </div>
);

export default Reviews;
