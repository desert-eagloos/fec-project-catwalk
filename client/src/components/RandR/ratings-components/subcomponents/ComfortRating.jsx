import React from 'react';
import { Row, Col } from 'react-bootstrap';
import PropTypes from 'prop-types';

export default function ComfortRating({ comfortValue }) {
  const value = ((comfortValue / 5) * 100) ?? 50;

  return (
    <>
      <Row className="m-0 p-0">
        Size
      </Row>
      <Row className="m-0 p-0">
        <span style={{ marginLeft: `${value}%` }}><i className="bi bi-caret-down-fill" /></span>
      </Row>
      <Row className="m-0 p-0">
        <Col className="m-0 p-0 d-flex justify-content-start"><span>Poor</span></Col>
        <Col className="m-0 p-0 d-flex justify-content-end"><span>Perfect</span></Col>
      </Row>
    </>
  );
}

ComfortRating.defaultProps = {
  comfortValue: 3,
};

ComfortRating.propTypes = {
  comfortValue: PropTypes.number,
};
