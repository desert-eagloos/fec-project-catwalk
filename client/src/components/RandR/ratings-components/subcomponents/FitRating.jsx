import React from 'react';
import { Row, Col } from 'react-bootstrap';
import PropTypes from 'prop-types';

export default function FitRating({ fitValue }) {
  const value = ((fitValue / 5) * 100) ?? 50;

  return (
    <>
      <Row className="m-0 p-0">
        Size
      </Row>
      <Row className="m-0 p-0">
        <span style={{ marginLeft: `${value}%` }}><i className="bi bi-caret-down-fill" /></span>
      </Row>
      <Row className="m-0 p-0">
        <Col className="m-0 p-0 d-flex justify-content-start"><span>Too Small</span></Col>
        <Col className="m-0 p-0 d-flex justify-content-center"><span>Perfect</span></Col>
        <Col className="m-0 p-0 d-flex justify-content-end"><span>Too Large</span></Col>
      </Row>
    </>
  );
}

FitRating.defaultProps = {
  fitValue: '3',
};

FitRating.propTypes = {
  fitValue: PropTypes.string,
};
