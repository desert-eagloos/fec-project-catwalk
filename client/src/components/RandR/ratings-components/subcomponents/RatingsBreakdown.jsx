import React from 'react';
import { Row, Col, ProgressBar } from 'react-bootstrap';
import PropTypes from 'prop-types';

import '../../../../css/RandR/Ratings/RatingBreakdown.css';

export default function RatingsBreakdown({ ratings }) {
  const totalRatings = Object.values(ratings)
    .reduce((sum, x) => sum + Number(x), 0);

  const rand = () => Math.floor(Math.random() * 1);

  return (
    Object.entries(ratings)
      .sort((a, b) => b[0].localeCompare(a[0]))
      .map(([rating, value], i) => (
        <Row key={`row${i + rand()}`} className="m-0 p-0">
          <Col sm={2} className="m-0 p-0">
            {rating}
          </Col>
          <Col className="m-0 p-0">
            <ProgressBar now={(value / totalRatings) * 100} />
          </Col>
        </Row>
      ))
  );
}

RatingsBreakdown.defaultProps = {
  ratings: {
    5: 1,
    4: 3,
    3: 2,
    2: 1,
    1: 0
  },
};

RatingsBreakdown.propTypes = {
  ratings: PropTypes.shape({
    5: PropTypes.number,
    4: PropTypes.number,
    3: PropTypes.number,
    2: PropTypes.number,
    1: PropTypes.number,
  }),
};
