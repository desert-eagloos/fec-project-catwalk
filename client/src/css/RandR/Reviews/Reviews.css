.randr-reviews-container {
  border: 2px solid #17a59f;
  border-radius: 2px;
  grid-area: reviews;
  display: grid;
  grid-template-columns: 1fr;
  grid-template-areas: 
  "total total"
  "ratinguserdate ratinguserdate"
  "reviewtitle reviewtitle"
  "reviewbody reviewbody"
  "helpfulreport helpfulreport"
  "buttons buttons";
}

.randr-reviews-container > div {
  border: 2px solid #17a59f;
  border-radius: 2px;
}

.randr-reviews-container > div:nth-child(1) {
  grid-area: total;
}

.randr-reviews-container > div:nth-child(2) {
  grid-area: ratinguserdate;
  display: flex;
  justify-content: space-between;
}

.randr-reviews-container > div:nth-child(3) {
  grid-area: reviewtitle;
}

.randr-reviews-container > div:nth-child(4) {
  grid-area: reviewbody;
}

.randr-reviews-container > div:nth-child(5) {
  grid-area: helpfulreport;
}

.randr-reviews-container > div:nth-child(6) {
  grid-area: buttons;
  display: flex;
  column-gap: 1rem;
}
