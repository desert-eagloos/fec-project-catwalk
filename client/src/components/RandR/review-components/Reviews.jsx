import React, { useState, useEffect } from 'react';
import axios from 'axios';
import {
  Container, Row, Col, Button, Dropdown, DropdownButton, Modal, Form,
} from 'react-bootstrap';
import Rating from 'react-rating';
import PropTypes from 'prop-types';
import _ from 'underscore';

import { roundToNearestQuarter } from '../../../utils/ratings';

import '../../../css/RandR/Reviews/Reviews.css';
import SizeRatingRadio from './SizeRatingRadio';
import WidthRatingRadio from './WidthRatingRadio';
import ComfortRatingRadio from './ComfortRatingRadio';
import LengthRatingRadio from './LengthRatingRadio';
import QualityRatingRadio from './QualityRatingRadio';
import FitRatingRadio from './FitRatingRadio';

const Reviews = ({ productId }) => {
  const [allReviews, setAllReviews] = useState([]);
  const [sortReviewBy, setSortReviewBy] = useState('relevant');
  const [isLoading, setIsLoading] = useState(true);
  const [reviewCount, setReviewCount] = useState(2);
  const [showMoreReviews, setShowMoreReviews] = useState(true);
  const [showMoreReviewsButton, setShowMoreReviewsButton] = useState(true);
  const [showAddReviewForm, setShowAddReviewForm] = useState(false);
  const [productCharacteristics, setProductCharacteristics] = useState({});

  const handleShowAddReviewForm = () => setShowAddReviewForm(true);
  const handleCloseAddReviewForm = () => setShowAddReviewForm(false);
  const handleSubmitReview = () => 'here';

  const getReviewsFromAPI = async (page, count, sort) => {
    const response = await axios.get('/reviews', {
      params: {
        productId, page, count, sort,
      },
    });
    return response.data;
  };

  const getAllReviewsFromAPI = async (sortBy = 'relevant') => {
    const getPartialReviews = async (count = 10) => {
      const reviews = await getReviewsFromAPI(1, count, sortBy);

      if (reviews.results.length < count) {
        return reviews.results;
      }
      const thisCount = count * 2;
      return getPartialReviews(thisCount);
    };
    return getPartialReviews();
  };

  const getReviews = () => allReviews.slice(0, reviewCount);

  const getProductCharacteristics = async () => {
    const result = await axios.get(`/reviews/meta?productId=${productId}`);
    const chars = result.data.characteristics;
    setProductCharacteristics(_.map(chars, (char, key) => {
      const prodRadio = [];
      switch (key) {
        case 'Comfort':
          prodRadio.push(<ComfortRatingRadio key="cfr" />);
          break;
        case 'Size':
          prodRadio.push(<SizeRatingRadio key="srr" />);
          break;
        case 'Fit':
          prodRadio.push(<FitRatingRadio key="frr" />);
          break;
        case 'Width':
          prodRadio.push(<WidthRatingRadio key="wrr" />);
          break;
        case 'Quality':
          prodRadio.push(<QualityRatingRadio key="qrr" />);
          break;
        case 'Length':
          prodRadio.push(<LengthRatingRadio key="lrr" />);
          break;
        default:
      }
      return prodRadio;
    }));
  };

  const rand = () => Math.floor(Math.random() * 1);

  const renderReviews = () => {
    const reviews = getReviews();
    return (reviews.map((review, index) => (
      <div key={`review${index + rand()}`}>
        <Row>
          <Col>
            <Rating
              key={`rating${index + rand()}`}
              emptySymbol="fa fa-star-o"
              fullSymbol="fa fa-star"
              initialRating={roundToNearestQuarter(review.rating)}
              fractions={4}
              readonly
            />
          </Col>
          <Col>
            {review.reviewer_name.toLowerCase()}
            ,
            {' '}
            {new Date(review.date).toDateString()}
          </Col>
        </Row>
        <Row>
          {review.summary.length > 60 ? `${review.summary.substring(0, 60)}...` : review.summary}
          {' '}
        </Row>
        <Row>{review.summary.length > 60 ? `${review.summary.substring(61)} \n\n ${review.body}` : review.body}</Row>
        <Row>
          <Col>
            Helpful? Yes &nbsp;
            {review.helpfulness}
          </Col>
          <Col>| Report</Col>
        </Row>
      </div>
    ))
    );
  };

  const renderReviewTotalAndSort = () => (
    <Row>
      {allReviews.length}
      {' '}
      reviews, sorted by &nbsp;
      <DropdownButton id="sortBy" title={sortReviewBy}>
        <Dropdown.Item key="1" eventKey="relevance" onSelect={(eventKey) => { setSortReviewBy(eventKey); }} active>relevance</Dropdown.Item>
        <Dropdown.Item key="2" eventKey="newest" onSelect={(eventKey) => { setSortReviewBy(eventKey); }}>newest</Dropdown.Item>
        <Dropdown.Item key="3" eventKey="helpfulness" onSelect={(eventKey) => { setSortReviewBy(eventKey); }}>helpfulness</Dropdown.Item>
      </DropdownButton>
    </Row>
  );

  const renderReviewButtons = () => (
    <>
      <Col>
        {
          showMoreReviews
            ? (
              <Button onClick={() => {
                setReviewCount(reviewCount + 2);
                setShowMoreReviewsButton(!showMoreReviewsButton);
              }}
              >
                More Review
              </Button>
            )
            : ''
        }
      </Col>
      <Col>
        <Button onClick={handleShowAddReviewForm}>Add a Review</Button>
        <Modal show={showAddReviewForm} onHide={handleCloseAddReviewForm} backdrop="static" keyboard={false} centered size="xl">
          <Container className="p-4">
            <Modal.Header closeButton>
              <Modal.Title>Add a Review</Modal.Title>
            </Modal.Header>
            <Modal.Body>
              <fieldset>
                <Form>
                  <Container>
                    <Form.Group as={Row} controlId="formStarRating">
                      <fieldset>
                        <Col>
                          <Form.Label as="legend">Star Rating</Form.Label>
                        </Col>
                        <Col>
                          <Rating
                            emptySymbol="fa fa-star-o"
                            fullSymbol="fa fa-star"
                            initialRating={0}
                          />
                        </Col>
                      </fieldset>
                    </Form.Group>
                    <Form.Group as={Row} controlId="formRecommendation">
                      <fieldset>
                        <Col>
                          <Form.Label as="legend">Do you recommend this product?</Form.Label>
                        </Col>
                        <Col>
                          <Form.Check inline label="Yes" type="radio" name="recommended" id="recommend-1" value="true" />
                          <Form.Check inline label="No" type="radio" name="recommended" id="recommend-2" value="false" />
                        </Col>
                      </fieldset>
                    </Form.Group>
                    <fieldset>
                      <Form.Group as={Row} controlId="formCharacteristics">
                        <Col className="mt-3">
                          <Form.Label as="legend">Characteristics</Form.Label>
                        </Col>
                      </Form.Group>
                      <Row className="justify-content-center">
                        {
                          _.map(productCharacteristics, (chars) => chars)
                        }
                      </Row>
                    </fieldset>
                  </Container>
                </Form>
              </fieldset>
            </Modal.Body>
            <Modal.Footer>
              <Button onClick={handleCloseAddReviewForm}>Cancel</Button>
              <Button onClick={handleSubmitReview}>Submit Review</Button>
            </Modal.Footer>
          </Container>
        </Modal>
      </Col>
    </>
  );

  const renderReviewsComponent = () => (
    <Container>
      <Row>
        {' '}
        {renderReviewTotalAndSort()}
        {' '}
      </Row>
      <Row>
        {' '}
        {renderReviews()}
        {' '}
      </Row>
      <Row>
        {' '}
        {renderReviewButtons()}
        {' '}
      </Row>
    </Container>
  );

  useEffect(async () => {
    const unsortedReviews = await getAllReviewsFromAPI();
    setAllReviews(unsortedReviews);
    setShowMoreReviewsButton(!showMoreReviewsButton);
    getProductCharacteristics();
    setIsLoading(false);
  }, []);

  useEffect(async () => {
    const sortedReviews = await getAllReviewsFromAPI(sortReviewBy);
    setAllReviews(sortedReviews);
    renderReviewsComponent();
  }, [sortReviewBy]);

  useEffect(() => {
    if (reviewCount < allReviews.length) {
      setShowMoreReviews(true);
    } else {
      setShowMoreReviews(false);
    }
  }, [showMoreReviewsButton]);

  if (isLoading) {
    return 'Reviews are loading...';
  }

  return (
    <>
      { renderReviewsComponent()}
    </>
  );
};

Reviews.propTypes = {
  productId: PropTypes.number,
};

Reviews.defaultProps = {
  productId: 18201,
};

export default Reviews;
