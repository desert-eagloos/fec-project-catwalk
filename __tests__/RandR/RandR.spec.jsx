import React from 'react';
import renderer from 'react-test-renderer';

import RatingsBreakdown from '../../client/src/components/RandR/ratings-components/subcomponents/RatingsBreakdown';
import FitRating from '../../client/src/components/RandR/ratings-components/subcomponents/FitRating';
import ComfortRating from '../../client/src/components/RandR/ratings-components/subcomponents/ComfortRating';
import Rating from 'react-rating';
import Reviews from '../../client/src/components/RandR/review-components/Reviews';

describe('Creates Ratings and Reviews Components', () => {
  describe('Creates a Star Rating Component', () => {
    test('Star', () => {
      const component = renderer.create(<Rating />);
      const json = component.toJSON();
      expect(json).toMatchSnapshot();
    });
  });
  describe('Creates a Rating Breakdown Component', () => {
    test(' Rating Breakdown', () => {
      const ratings = {
        5: 2,
        4: 3,
        3: 5,
        2: 1,
        1: 10,
      };
      const component = renderer.create(<RatingsBreakdown key="bcTest1" ratings={ratings} />);
      const json = component.toJSON();
      expect(json).toMatchSnapshot();
    });
  });
  describe('Creates a Size Fit Rating Component', () => {
    test('Size Fit Rating', () => {
      const component = renderer.create(<FitRating ratings={3} />);
      const json = component.toJSON();
      expect(json).toMatchSnapshot();
    });
  });
  describe('Creates a Comfort Fit Rating Component', () => {
    test('Comfort Fit Rating', () => {
      const component = renderer.create(<ComfortRating ratings={2} />);
      const json = component.toJSON();
      expect(json).toMatchSnapshot();
    });
  });
  describe('Creates a Review Component', () => {
    test('Reviews', () => {
      const component = renderer.create(<Reviews />);
      const json = component.toJSON();
      expect(json).toMatchSnapshot();
    });
  });
});
