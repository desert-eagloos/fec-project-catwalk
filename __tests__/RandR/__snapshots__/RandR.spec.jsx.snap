// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creates Ratings and Reviews Components Creates a Comfort Fit Rating Component Comfort Fit Rating 1`] = `
Array [
  <div
    className="m-0 p-0 row"
  >
    Size
  </div>,
  <div
    className="m-0 p-0 row"
  >
    <span
      style={
        Object {
          "marginLeft": "60%",
        }
      }
    >
      <i
        className="bi bi-caret-down-fill"
      />
    </span>
  </div>,
  <div
    className="m-0 p-0 row"
  >
    <div
      className="m-0 p-0 d-flex justify-content-start col"
    >
      <span>
        Poor
      </span>
    </div>
    <div
      className="m-0 p-0 d-flex justify-content-end col"
    >
      <span>
        Perfect
      </span>
    </div>
  </div>,
]
`;

exports[`Creates Ratings and Reviews Components Creates a Rating Breakdown Component  Rating Breakdown 1`] = `
Array [
  <div
    className="m-0 p-0 row"
  >
    <div
      className="m-0 p-0 col-sm-2"
    >
      5
    </div>
    <div
      className="m-0 p-0 col"
    >
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={9.523809523809524}
          className="progress-bar"
          role="progressbar"
          style={
            Object {
              "width": "9.524%",
            }
          }
        />
      </div>
    </div>
  </div>,
  <div
    className="m-0 p-0 row"
  >
    <div
      className="m-0 p-0 col-sm-2"
    >
      4
    </div>
    <div
      className="m-0 p-0 col"
    >
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={14.285714285714285}
          className="progress-bar"
          role="progressbar"
          style={
            Object {
              "width": "14.286%",
            }
          }
        />
      </div>
    </div>
  </div>,
  <div
    className="m-0 p-0 row"
  >
    <div
      className="m-0 p-0 col-sm-2"
    >
      3
    </div>
    <div
      className="m-0 p-0 col"
    >
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={23.809523809523807}
          className="progress-bar"
          role="progressbar"
          style={
            Object {
              "width": "23.81%",
            }
          }
        />
      </div>
    </div>
  </div>,
  <div
    className="m-0 p-0 row"
  >
    <div
      className="m-0 p-0 col-sm-2"
    >
      2
    </div>
    <div
      className="m-0 p-0 col"
    >
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={4.761904761904762}
          className="progress-bar"
          role="progressbar"
          style={
            Object {
              "width": "4.762%",
            }
          }
        />
      </div>
    </div>
  </div>,
  <div
    className="m-0 p-0 row"
  >
    <div
      className="m-0 p-0 col-sm-2"
    >
      1
    </div>
    <div
      className="m-0 p-0 col"
    >
      <div
        className="progress"
      >
        <div
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={47.61904761904761}
          className="progress-bar"
          role="progressbar"
          style={
            Object {
              "width": "47.619%",
            }
          }
        />
      </div>
    </div>
  </div>,
]
`;

exports[`Creates Ratings and Reviews Components Creates a Review Component Reviews 1`] = `"Reviews are loading..."`;

exports[`Creates Ratings and Reviews Components Creates a Size Fit Rating Component Size Fit Rating 1`] = `
Array [
  <div
    className="m-0 p-0 row"
  >
    Size
  </div>,
  <div
    className="m-0 p-0 row"
  >
    <span
      style={
        Object {
          "marginLeft": "60%",
        }
      }
    >
      <i
        className="bi bi-caret-down-fill"
      />
    </span>
  </div>,
  <div
    className="m-0 p-0 row"
  >
    <div
      className="m-0 p-0 d-flex justify-content-start col"
    >
      <span>
        Too Small
      </span>
    </div>
    <div
      className="m-0 p-0 d-flex justify-content-center col"
    >
      <span>
        Perfect
      </span>
    </div>
    <div
      className="m-0 p-0 d-flex justify-content-end col"
    >
      <span>
        Too Large
      </span>
    </div>
  </div>,
]
`;

exports[`Creates Ratings and Reviews Components Creates a Star Rating Component Star 1`] = `
<span
  onMouseLeave={[Function]}
  style={
    Object {
      "direction": "ltr",
      "display": "inline-block",
    }
  }
>
  <span
    onClick={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "display": "inline-block",
        "position": "relative",
      }
    }
  >
    <span
      style={Object {}}
    >
      <span
        style={
          Object {
            "backgroundColor": "#ccc",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
    <span
      style={
        Object {
          "display": "inline-block",
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "top": 0,
          "width": "0%",
        }
      }
    >
      <span
        style={
          Object {
            "backgroundColor": "black",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
  </span>
  <span
    onClick={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "display": "inline-block",
        "position": "relative",
      }
    }
  >
    <span
      style={Object {}}
    >
      <span
        style={
          Object {
            "backgroundColor": "#ccc",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
    <span
      style={
        Object {
          "display": "inline-block",
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "top": 0,
          "width": "0%",
        }
      }
    >
      <span
        style={
          Object {
            "backgroundColor": "black",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
  </span>
  <span
    onClick={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "display": "inline-block",
        "position": "relative",
      }
    }
  >
    <span
      style={Object {}}
    >
      <span
        style={
          Object {
            "backgroundColor": "#ccc",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
    <span
      style={
        Object {
          "display": "inline-block",
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "top": 0,
          "width": "0%",
        }
      }
    >
      <span
        style={
          Object {
            "backgroundColor": "black",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
  </span>
  <span
    onClick={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "display": "inline-block",
        "position": "relative",
      }
    }
  >
    <span
      style={Object {}}
    >
      <span
        style={
          Object {
            "backgroundColor": "#ccc",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
    <span
      style={
        Object {
          "display": "inline-block",
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "top": 0,
          "width": "0%",
        }
      }
    >
      <span
        style={
          Object {
            "backgroundColor": "black",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
  </span>
  <span
    onClick={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "display": "inline-block",
        "position": "relative",
      }
    }
  >
    <span
      style={Object {}}
    >
      <span
        style={
          Object {
            "backgroundColor": "#ccc",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
    <span
      style={
        Object {
          "display": "inline-block",
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "top": 0,
          "width": "0%",
        }
      }
    >
      <span
        style={
          Object {
            "backgroundColor": "black",
            "border": "5px double white",
            "borderRadius": "50%",
            "display": "inline-block",
            "height": 30,
            "width": 30,
          }
        }
      />
    </span>
  </span>
</span>
`;
